execution_engine/src/core/runtime/mod.rs

```diff
@@ -2275,43 +2331,55 @@ where
         }
     }
 
-    /// Transfers `amount` of motes from `source` purse to `target` purse.
-    #[allow(clippy::too_many_arguments)]
-    fn transfer_from_purse_to_purse(
+    fn transfer_from_purse_to_account(
         &mut self,
-        source_ptr: u32,
-        source_size: u32,
-        target_ptr: u32,
-        target_size: u32,
-        amount_ptr: u32,
-        amount_size: u32,
-        id_ptr: u32,
-        id_size: u32,
-    ) -> Result<Result<(), mint::Error>, Error> {
-        let source: URef = {
-            let bytes = self.bytes_from_mem(source_ptr, source_size as usize)?;
-            bytesrepr::deserialize(bytes).map_err(Error::BytesRepr)?
-        };
+        source: URef,
+        target_account: &Account,
+        amount: U512,
+        id: Option<u64>,
+    ) -> Result<TransferResult, Error> {
+        // Attenuate the target main purse
+        let target_uref = target_account.main_purse_add_only();
 
-        let target: URef = {
-            let bytes = self.bytes_from_mem(target_ptr, target_size as usize)?;
-            bytesrepr::deserialize(bytes).map_err(Error::BytesRepr)?
-        };
+        if source.with_access_rights(AccessRights::ADD) == target_uref {
+            return Ok(Ok(TransferredTo::ExistingAccount));
+        }
 
-        let amount: U512 = {
-            let bytes = self.bytes_from_mem(amount_ptr, amount_size as usize)?;
-            bytesrepr::deserialize(bytes).map_err(Error::BytesRepr)?
-        };
+        // Upsert ADD access to caller on target allowing deposit of motes; this will be
```

Comment:
Did you mean?
```suggestion
        // Insert ADD access to caller on target allowing deposit of motes; this will be
```

Or maybe just say `Grant ADD access...`.

Reply:
Reworded to `Grant`